packages/datadog-plugin-playwright/src/index.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const { storage } = require('../../datadog-core')
+const CiPlugin = require('../../dd-trace/src/plugins/ci_plugin')
+
+const {
+  TEST_STATUS,
+  finishAllTraceSpans,
+  getTestParentSpan,
+  getTestSessionCommonTags,
+  getTestSuitePath,
+  TEST_SESSION_ID,
+  TEST_COMMAND,
+  TEST_SUITE_ID,
+  getTestSuiteCommonTags
+} = require('../../dd-trace/src/plugins/util/test')
+const { RESOURCE_NAME } = require('../../../ext/tags')
+const { COMPONENT } = require('../../dd-trace/src/constants')
+
+class PlaywrightPlugin extends CiPlugin {
+  static get name () {
+    return 'playwright'
+  }
+
+  constructor (...args) {
+    super(...args)
+
+    this._testSuites = new Map()
+
+    this.addSub('ci:playwright:session:start', ({ command, frameworkVersion, rootDir }) => {
+      const childOf = getTestParentSpan(this.tracer)
+      this.command = command
+      this.frameworkVersion = frameworkVersion
+      this.rootDir = rootDir
+
+      const testSessionSpanMetadata = getTestSessionCommonTags(command, frameworkVersion)
+      this.testSessionSpan = this.tracer.startSpan('playwright.test_session', {
+        childOf,
+        tags: {
+          [COMPONENT]: this.constructor.name,
+          ...this.testEnvironmentMetadata,
+          ...testSessionSpanMetadata
+        }
+      })
+    })
+
+    this.addSub('ci:playwright:session:finish', ({ status, onDone }) => {
+      this.testSessionSpan.setTag(TEST_STATUS, status)
+      this.testSessionSpan.finish()
+      finishAllTraceSpans(this.testSessionSpan)
+      this.tracer._exporter.flush(onDone)
+    })
+
+    this.addSub('ci:playwright:test-suite:start', (testSuiteAbsolutePath) => {
+      const store = storage.getStore()
+      const testSuite = getTestSuitePath(testSuiteAbsolutePath, this.rootDir)
+
+      const testSuiteMetadata = getTestSuiteCommonTags(
+        this.command,
+        this.frameworkVersion,
+        testSuite
+      )
+
+      const testSuiteSpan = this.tracer.startSpan('playwright.test_suite', {
+        childOf: this.testSessionSpan,
+        tags: {
+          [COMPONENT]: this.constructor.name,
+          ...this.testEnvironmentMetadata,
+          ...testSuiteMetadata
+        }
+      })
+      this.enter(testSuiteSpan, store)
+
+      this._testSuites.set(testSuite, testSuiteSpan)
+    })
+
+    this.addSub('ci:playwright:test-suite:finish', (status) => {
+      const span = storage.getStore().span
+      span.setTag(TEST_STATUS, status)
+      span.finish()
+    })
+
+    this.addSub('ci:playwright:test:start', ({ testName, testSuiteAbsolutePath }) => {
+      const store = storage.getStore()
+      const testSuite = getTestSuitePath(testSuiteAbsolutePath, this.rootDir)
+      const span = this.startTestSpan(testName, testSuite)
+
+      this.enter(span, store)
+    })
+    this.addSub('ci:playwright:test:finish', ({ testStatus, steps, error }) => {
+      const span = storage.getStore().span
```

Comment:
Probably shouldn't just assume a span was found. Context loss could happen, we should code in a way that avoids the process crashing when it does.

```suggestion
      const store = storage.getStore()
      const span = store && store.span
      if (!span) return
```

Might also be worth copying the `activeSpan` getter from TracingPlugin into CiPlugin or maybe even the Plugin base class if you're going to need this pattern a bunch.

Reply:
good idea, thanks. I've added it just here but I'll see if I can do the getter in another PR 👍